src/util/cilfacade.ml

```diff
@@ -80,6 +80,109 @@ let end_basic_blocks f =
   let thisVisitor = new allBBVisitor in
   visitCilFileSameGlobals thisVisitor f
 
+(* replaces goto s with goto s' when newtarget s = Some s' IN PLACE *)
+class patchLabelsGotosVisitor(newtarget) = object
+  inherit nopCilVisitor
+
+  method! vstmt s =
+    match s.skind with
+    | Goto (target,loc) ->
+      (match newtarget target with
+       | None -> SkipChildren
+       | Some nt -> s.skind <- Goto (nt, loc); DoChildren)
+    | _ -> DoChildren
+end
+
+(*
+  Makes a copy, replacing top-level breaks with goto loopEnd and top-level continues with
+  goto currentIterationEnd
+  Also assigns fresh names to all labels and patches gotos for labels appearing in the current
+  fragment to their new name
+*)
+class copyandPatchLabelsVisitor(loopEnd,currentIterationEnd) = object
+  inherit nopCilVisitor
+
+  val mutable depth = 0
+  val mutable loopNestingDepth = 0
+  val gotos = Hashtbl.create 20
+
+  method! vstmt s =
+    let after x =
+      depth <- depth-1;
+      if depth = 0 then
+        (* the labels can only be patched once the entire part of the AST we want has been transformed, and *)
+        (* we know all lables appear in the hash table *)
+        let patchLabelsVisitor = new patchLabelsGotosVisitor(Hashtbl.find_opt gotos) in
+        let x  = visitCilStmt patchLabelsVisitor x in
+        Hashtbl.clear gotos;
+        x
+      else
+        x
+    in
+    let rename_labels sn =
+      let new_labels = List.map (function Label(str,loc,b) -> Label (Cil.freshLabel str,loc,b) | x -> x) sn.labels in
+      (* this makes new physical copy*)
+      let new_s = {sn with labels = new_labels} in
+      if new_s.labels <> [] then
+        (* Use original s, ns might be temporay e.g. if the type of statement changed *)
+        (* record that goto s; appearing in the current fragment should later be patched to goto new_s *)
+        Hashtbl.add gotos (ref s) (ref new_s);
```

Comment:
The keys of the hashtable are _mutable_ references? That's a big no, because if such reference actually changes, then the (polymorphic) hash for it could change and break the hashtable lookups.

Reply:
We have the guarantee that these do not change during this phase though (as it would break invariants in a lot of other places). I don't think there is anything else one can use to uniquely identify statements.
In principle one could make an association list based on `==` of `stmt`, but that would be linear in the lookup, so I'm not sure that this is much better.

Reply:
I might try going via the names then maybe, somehow feels a bit less elegant. Any other suggestions?

Reply:
If they need to be identified by physical equality, then a `Hashtbl.Make` with the following in the argument module would make that more explicit:
https://github.com/goblint/analyzer/blob/2669f675b7f64f92e1f5db876c429b68487ec67e/src/framework/cfgTools.ml#L59-L61
Of course that still assumes for `hash` that the fields aren't mutated, but at least there isn't an additional wrapper reference to mutate.

Reply:
:+1: